### PR TITLE
Faster and deeper pyramid retaining

### LINKED ISSUE #1570: when zooming out, map flashes blank
This doesn't happen when zooming in, only when zooming out:

Might be the same as https://github.com/mapbox/mapbox-gl-js/issues/934

### PR DESCRIPTION
**Pull request by mourner**
Bump children retaining to +3 zooms and implement a blazing fast `findLoadedChildren` routine. Profiling `flyTo` from DC to SF, `findLoadedChildren` takes 3.7ms. Closes #1570. cc @jfirebaugh @bhousel

### CODE REVIEW

**Comment by mourner** [comment]
Needs a port to native as well, should be easy.

**Comment by jfirebaugh** [comment]
@mourner Can you explain the algorithm? I'm having trouble seeing how it works. It looks like for the input `coord`, only `coord.z` is used -- can this be correct?

**Comment by mourner** [comment]
You're right, we should only consider tiles that are children of the current tile. Will fix now.

**Comment by mourner** [comment]
@jfirebaugh Fixed. The following line can look a bit weird, but that's the only way around the fact that tile coordinates are not consistent — if a tile is overscaled, the coord stores overscaled zoom but x/y coordinates as if it was at maxzoom.

```js
var z2 = Math.pow(2, Math.min(tile.coord.z, this.maxzoom) - Math.min(coord.z, this.maxzoom));
```

File: js/source/tile_pyramid.js [hunk 1]
```diff
@@ -146,22 +146,36 @@ TilePyramid.prototype = {
      * @private
      */
     findLoadedChildren: function(coord, maxCoveringZoom, retain) {
```
Comment (jfirebaugh) [inline js/source/tile_pyramid.js]: `maxCoveringZoom` is no longer used; is this intentional?

File: js/source/tile_pyramid.js [hunk 2]
```diff
@@ -146,22 +146,36 @@ TilePyramid.prototype = {
...
+                tile = this._tiles[parentId];
+
+                if (tile && tile.loaded) {
+                    delete retain[id];
+                    retain[parentId] = true;
                 }
             }
         }
-        return complete;
+        return found;
```
Comment (jfirebaugh) [inline js/source/tile_pyramid.js]: This changes the sense of the return value: before it was "did we find children that completely cover the parent tile", now it's "did we find one or more children". The return value is used in both contexts where `findLoadedChildren` is called, but I'm not sure how important the existing semantics are. What do you think?

Comment (mourner) [inline js/source/tile_pyramid.js]: In theory, the old approach can lead to better coverage, but in practice, the difference is only apparent in situation where a tile is not loaded but both its parent and some children are loaded, which is extremely rare. So it's probably not important.

File: js/source/tile_pyramid.js (hunk 1 above)
Comment (mourner) [inline js/source/tile_pyramid.js]: Not intentional, I forgot about it completely. :) Will add a check.

**Comment by mourner** [comment]
@jfirebaugh fixed maxCoveringZoom.

**Comment by jfirebaugh** [comment]
:+1:

**Pull request merged**